state/protocol/protocol_state/updater.go

```diff
@@ -62,99 +61,75 @@ func (u *Updater) Build() (updatedState *flow.ProtocolStateEntry, stateID flow.I
-	// - For the next epoch we need to additionally include identities from previous(current) epoch with 0 weight.
-	// We will do this in next steps:
-	// 1. Add identities from current epoch setup event to currentEpochIdentities.
-	// 2. Add identities from next epoch setup event to currentEpochIdentities, with 0 weight,
-	//     but only if they are not present in currentEpochIdentities.
-	// 3. Add identities from next epoch setup event to nextEpochIdentities.
-	// 4. Add identities from current epoch setup event to nextEpochIdentities, with 0 weight,
-	//    but only if they are not present in nextEpochIdentities.
 
-	// lookup of dynamic data for current protocol state identities
-	// by definition, this will include identities from current epoch + identities from previous epoch with 0 weight.
-	activeIdentitiesLookup := u.parentState.CurrentEpoch.ActiveIdentities.Lookup()
+	// When observing setup event for subsequent epoch, construct the EpochStateContainer for `ProtocolStateEntry.NextEpoch`.
+	// Context:
+	// Note that the `EpochStateContainer.ActiveIdentities` only contains the nodes that are *active* in the next epoch. Active means
+	// that these nodes are authorized to contribute to extending the chain. Nodes are listed in `ActiveIdentities` if and only if
+	// they are part of the EpochSetup event for the respective epoch.
+	//
+	// sanity checking SAFETY-CRITICAL INVARIANT (I):
+	// While ejection status and dynamic weight are not part of the EpochSetup event, we can supplement this information as follows:
+	//   - Per convention, service events are delivered (asynchronously) in an *order-preserving* manner. Furthermore, weight changes or
+	//     node ejection is entirely mediated by system smart contracts and delivered via service events.
+	//   - Therefore, the EpochSetup event contains the up-to-date snapshot of the cluster members. Any weight changes or node ejection
+	//     that happened before should be reflected in the EpochSetup event. Specifically, the initial weight should be reduced and ejected
+	//     nodes should be no longer listed in the EpochSetup event.
+	//   - Hence, the following invariant must be satisfied by the system smart contracts for all active nodes in the upcoming epoch:
+	//      (i) When the EpochSetup event is emitted / processed, the weight of all active nodes equals their InitialWeight and
+	//     (ii) The Ejected flag is false. Node X being ejected in epoch N (necessarily via a service event emitted by the system
+	//          smart contracts earlier) but also being listed in the setup event for the subsequent epoch (service event emitted by
+	//          the system smart contracts later) is illegal.
+	// sanity checking SAFETY-CRITICAL INVARIANT (II):
+	//   - Per convention, the system smart contracts should list the IdentitySkeletons in canonical order. This is useful for
+	//     most efficient construction of the full active Identities for an epoch.
```

Comment:
Can we update our smart contract to do this? Either way we would need to work on it in future

Reply:
I don't think we should have the smart contract do sorting. It's a resource-constrained environment and it is not critical that the set be sorted within the system contract.

It seems fine to me for the protocol layer to translate the event into a representation that is easier for it to work with.

Reply:
Thanks Jordan for flagging that my comment in the code was incorrect. I have updated it in https://github.com/onflow/flow-go/pull/4834/commits/49934ce513f6f849a2c5ecb44a675ff45a610c58 to read: https://github.com/onflow/flow-go/blob/49934ce513f6f849a2c5ecb44a675ff45a610c58/state/protocol/protocol_state/updater.go#L92-L98

Thereby, the documentation precisely describes the current state, which I think is all that is needed for this PR (?) 

My thoughts on where _ordering could be implemented in the future_

1. On the one hand, I see (weak) benefits of not having an ordering requirement in the system smart contracts. Conceptually, a set is sufficient for the current functionality of the system smart contracts. _Not_  requiring an ordering is one less details that engineers have know about and that could accidentally be broken. 

  For similar reasons, a block proposer is not _required_  to include seals in an ordered manner. Instead, we re-order them when ingesting the block, which improves modularity of the code by removing inter-dependencies.  

2. On the other hand, in the future, the system smart contracts will also need to have the canonical ordering. This is because we want to eventually implement performance-dependent rewards for consensus nodes, which in turn requires a smart contract to decode the signer indices for blocks to determine which node contributed to a QC (the basis for the subsequent reward payout at the end of the epoch). To decode the signer indices, the smart contract needs to know the ordering. 
   * We then have two choices: (i) the system smart contract maintains an un-ordered set and we re-order the identities at every block (a huge waste of computation) or (ii) the system smart maintains the identities in canonical ordering
   * Btw, note that to decode QC signer identities, we also need a notion of identities for each epoch. In other words, we would no longer be able top the current epoch's identity table at the end of the staking phase 😅

Reply:
> requires a smart contract to decode the signer indices for blocks to determine which node contributed to a QC

Or, we could implement an API on the injected `Block` Cadence type that does this?

Reply:
> could implement an API on the injected `Block` Cadence type that does this?

We could. At which point to we change the publisher of the data (system smart contract), vs layering on logic on the consumers (protocol state, dynamic reward's logic). If we have only one consumer, I don't feel it makes much of a difference because the ordering has to live either on the consumer or the publisher -- but putting the ordering into the consumer improves modularity. With two consumers, we are starting to duplicate the ordering logic ...

Reply:
Thanks for the insightful discussion on this point. I feel for the scope of this PR, the comment is addressed (?)